File: backend/services/pdf_service.py
import fitz  # pymupdf


def extract_text_from_pdf(file_bytes: bytes) -> str:
    doc = fitz.open(stream=file_bytes, filetype="pdf")
    pages = [page.get_text() for page in doc]
    doc.close()
    return "\n\n".join(pages)


def chunk_text(text: str, chunk_size: int = 800, overlap: int = 100) -> list[str]:
    paragraphs = [p.strip() for p in text.split("\n\n") if p.strip()]
    chunks: list[str] = []
    current = ""

    for para in paragraphs:
        if len(current) + len(para) + 2 <= chunk_size:
            current = (current + "\n\n" + para).strip()
        else:
            if current:
                chunks.append(current)
            # Start new chunk with overlap from tail of previous chunk
            tail = current[-overlap:] if len(current) > overlap else current
            current = (tail + "\n\n" + para).strip()

    if current.strip():
        chunks.append(current.strip())

    return chunks

File: backend-node/services/pdfService.js
import pdfParse from 'pdf-parse/lib/pdf-parse.js';

export async function extractTextFromPdf(buffer) {
  const data = await pdfParse(buffer);
  return data.text;
}

export function chunkText(text, chunkSize = 800, overlap = 100) {
  const paragraphs = text.split(/\n\n+/).map(p => p.trim()).filter(Boolean);
  const chunks = [];
  let current = '';

  for (const para of paragraphs) {
    if (current.length + para.length + 2 <= chunkSize) {
      current = current ? `${current}\n\n${para}` : para;
    } else {
      if (current) chunks.push(current);
      const tail = current.slice(-overlap);
      current = tail ? `${tail}\n\n${para}` : para;
    }
  }
  if (current.trim()) chunks.push(current.trim());
  return chunks;
}
